### PR TITLE
feat: always provide `context` argument

### PR DESCRIPTION
**Pull request by SimenB**
I need access to the `request` object inside our message handler.

### CODE REVIEW

File: src/methods/fastify-plugin.ts [hunk 1]
```diff
@@ -20,21 +20,23 @@ const pubSubFastifyPluginFn = async <Data, Context>(
     },
     async (req, reply) => {
       try {
-        const res = await handlePubSubMessage<Data, Context>({
+        const res = await handlePubSubMessage<Data, FastifyRequest>({
           parser,
           message: req.body.message,
           handler,
           parseJson,
-          // context: req,
```
Comment (SimenB) [inline src/methods/fastify-plugin.ts:28]: no idea why this was commented out

File: src/methods/fastify-plugin.ts [hunk 2]
```diff
@@ -1,16 +1,16 @@
-import { FastifyInstance } from 'fastify';
+import { TypeBoxTypeProvider } from '@fastify/type-provider-typebox';
+import { FastifyInstance, FastifyRequest } from 'fastify';
 import fp from 'fastify-plugin';
 import { handlePubSubMessage } from '../common';
-import { PubSubConfig, PubSubRequest, PubSubRequestType } from '../types';
+import { PubSubConfig, PubSubRequest } from '../types';
 
-const pubSubFastifyPluginFn = async <Data, Context>(
+const pubSubFastifyPluginFn = async <Data = unknown>(
```
Comment (SimenB) [inline src/methods/fastify-plugin.ts:7]: I don't understand why you take `Context` as a generic when you just pass `request`. (or rather, before this PR it was never passed at all)

Comment (SimenB) [inline src/methods/fastify-plugin.ts:7]: not defaulting to something was a breaking change to the types introduced in #142

Comment (SimenB) [inline src/methods/fastify-plugin.ts:7]: FWIW, these types are still kinda hard to work with, but I don't really know how to solve it cleanly. so I won't attempt to 😀

File: src/methods/cloud-functions.ts [hunk 3]
```diff
@@ -20,21 +20,24 @@ export function createPubSubCloudFunctions<Data = unknown, Context = unknown>(
 ): CloudFunctionFun {
   const { parseJson, onError, logger } = options;
   return async (req, res): Promise<void> => {
+    const context: Context = req.body;
```
Comment (SimenB) [inline src/methods/cloud-functions.ts:23]: not sure why this pass only the `body` instead of the entire request - but I don't wanna change it now

**Comment by runeh** [comment]
> I need access to the `request` object inside our message handler.

It hasn't been passed in, since we were thinking maybe it wouldn't be needed. Using it to get at a trace-id header or whatnot makes sense though.

There has been an idea for a `context` object to be similar to what you'd use in graphql resolvers. So you'd pass in a `createContext` function when creating the pubsub listener.

I still kinda thing that makes sense, and that function could get the request object. So you'd do something like.

```typescript
const server = createPubSubServer({
  createContext: (request) => ({traceId: request.headers.traceId ?? randomUuid()}),
  handler: ({message, data, context}) => {
    // use message data and context.traceId here 
  }
})
```

Thoughts? It shouldn't be a big job anyway, and the end result feels like it'd be nicer.

**Comment by SimenB** [comment]
If a `createContext` is created that'd work great as well 👍 Right now `context` is a generic that is sometimes passed, sometimes not and that you cannot type yourself. Some sort of `createContext` that this library just pass through soudns way better 😀

**Comment by thced** [comment]
I find this a good solution as well 👍🏻

**Comment by thced** [comment]
Can this be released, please?

**Review by cobraz — APPROVED** [review]
This looks good! I also think we should refactor some of the typing things.

Thanks! 🙏

**Pull request merged**